packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+                    e.setHeader("Duplicate package import `{}`",
+                                absl::StrJoin(imported.fullName, "::", NameFormatter(ctx)));
+                    e.addErrorLine(core::Loc(ctx.file, importedNames[imported.mangledName]),
+                                   "Previous package import found here");
+                }
+            } else {
+                importedNames[imported.mangledName] = imported.loc;
+                const auto &found = *it->second;
+
+                ast::ClassDef::RHS_store exportedItemsCopy;
+                for (const auto &exported : found.exportedItems) {
+                    exportedItemsCopy.emplace_back(exported.deepCopy());
+                }
+
+                ENFORCE(!imported.fullName.empty());
+                // Create a module for the imported package that sets up constant references to exported items.
+                // Use proper loc information on the module name so that `import Foo` displays in the results of LSP
+                // Find All References on `Foo`.
+                importedPackages.emplace_back(ast::MK::Module(imported.loc, core::Loc(ctx.file, imported.loc),
+                                                              imported.getFullNameLiteral(imported.loc), {},
+                                                              std::move(exportedItemsCopy)));
+            }
+        }
+    }
+
+    // Include Kernel in the package.
+    importedPackages.emplace_back(ast::MK::Send1(core::LocOffsets::none(), ast::MK::Self(core::LocOffsets::none()),
+                                                 core::Names::include(), name2Expr(core::Names::Constants::Kernel())));
+
+    auto packageNamespace =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(package.name.mangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(importedPackages));
+
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    rootKlass->rhs.emplace_back(move(packageNamespace));
+    return file;
+}
+
+ast::ParsedFile rewritePackagedFile(core::Context ctx, ast::ParsedFile file, core::NameRef packageMangledName) {
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    auto moduleWrapper =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(packageMangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(rootKlass->rhs));
+    rootKlass->rhs.clear();
+    rootKlass->rhs.emplace_back(move(moduleWrapper));
+    return file;
+}
+
+// We can't run packages without having all package ASTs. Assert that they are all present.
+bool checkContainsAllPackages(const core::GlobalState &gs, const vector<ast::ParsedFile> &files) {
```

Comment:
This function looks slow. What do you think about wrapping it in a `DEBUG_ONLY(...);` macro to make sure that it's only ever called from an ENFORCE?

Reply:
Ah that's doesn't work the way I thought it did. Never mind.

Reply:
It's only called from an `ENFORCE` right now and should never be called in a production build. I'm 100% OK with a `DEBUG_ONLY` wrapper -- I didn't know about it. :)

Reply:
We could potentially define `DEBUG_ONLY` in another way to get this to work:

```cpp
#ifdef DEBUG_MODE
#define DEBUG_ONLY(X) X
#else
#define DEBUG_ONLY(X)
#endif
```

but that's probably overkill. Up to you.

Reply:
I ended up making this method much cheaper so it's less of a concern.